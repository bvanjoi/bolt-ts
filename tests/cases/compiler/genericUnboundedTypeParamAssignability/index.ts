// From `github.com/microsoft/TypeScript/blob/v5.8.2/tests/cases/compiler/genericUnboundedTypeParamAssignability.ts`, Apache-2.0 License

function f1<T>(o: T) {
  o.toString(); // error
  //~^ ERROR: Property 'toString' does not exist on type 'T'.
}

function f2<T extends {}>(o: T) {
  o.toString(); // no error
}

function f3<T extends Record<string, any>>(o: T) {
  o.toString(); // no error
}

function user<T>(t: T) {
  f1(t);
  f2(t); // error in strict, unbounded T doesn't satisfy the constraint
  //~^ ERROR: Argument of type 'T' is not assignable to parameter of type '{ }'.
  f3(t); // error in strict, unbounded T doesn't satisfy the constraint
  //~^ ERROR: Argument of type 'T' is not assignable to parameter of type 'Record'.
  t.toString();  // error, for the same reason as f1()
  //~^ ERROR: Property 'toString' does not exist on type 'T'.
}

