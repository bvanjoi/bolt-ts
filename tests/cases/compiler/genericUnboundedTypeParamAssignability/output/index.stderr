  × Property 'toString' does not exist on type 'T'.
   ╭─[../../tests/cases/compiler/genericUnboundedTypeParamAssignability/index.ts:6:5]
 6 │ toString
   · ────────
   ╰────

  × Argument of type 'T' is not assignable to parameter of type '{ }'.
    ╭─[../../tests/cases/compiler/genericUnboundedTypeParamAssignability/index.ts:20:6]
 20 │ t
    · ─
    ╰────

  × Argument of type 'T' is not assignable to parameter of type 'Record'.
    ╭─[../../tests/cases/compiler/genericUnboundedTypeParamAssignability/index.ts:22:6]
 22 │ t
    · ─
    ╰────

  × Property 'toString' does not exist on type 'T'.
    ╭─[../../tests/cases/compiler/genericUnboundedTypeParamAssignability/index.ts:24:5]
 24 │ toString
    · ────────
    ╰────
